watertap/unit_models/boron_removal.py

```diff
@@ -0,0 +1,1123 @@
+            mol_Base = pyunits.convert(
+                self.caustic_cation_charge * self.caustic_dose[t] / self.caustic_mw,
+                to_units=units_meta("amount") * units_meta("length") ** -3,
+            )
+            return mol_H == mol_OH + mol_Borate + Alk - mol_Base
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="Total boron balance",
+        )
+        def eq_total_boron(self, t):
+            inlet_Boron = self.control_volume.properties_in[t].conc_mol_phase_comp[
+                "Liq", self.boron_name_id
+            ]
+            inlet_Borate = self.control_volume.properties_in[t].conc_mol_phase_comp[
+                "Liq", self.borate_name_id
+            ]
+            mol_Borate = pyunits.convert(
+                self.mol_Borate[t],
+                to_units=units_meta("amount") * units_meta("length") ** -3,
+            )
+            mol_Boron = pyunits.convert(
+                self.mol_Boron[t],
+                to_units=units_meta("amount") * units_meta("length") ** -3,
+            )
+            return inlet_Boron + inlet_Borate == mol_Borate + mol_Boron
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="Water dissociation",
+        )
+        def eq_water_dissociation(self, t):
+            return (
+                self.Kw_0
+                * exp(
+                    -self.dH_w
+                    / Constants.gas_constant
+                    / self.control_volume.properties_out[t].temperature
+                )
+            ) == self.mol_H[t] * self.mol_OH[t]
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="Boron dissociation",
+        )
+        def eq_boron_dissociation(self, t):
+            return (
+                self.Ka_0
+                * exp(
+                    -self.dH_a
+                    / Constants.gas_constant
+                    / self.control_volume.properties_out[t].temperature
+                )
+            ) * self.mol_Boron[t] == self.mol_H[t] * self.mol_Borate[t]
+
+        # Add constraints for mass transfer terms
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.config.property_package.phase_list,
+            self.config.property_package.component_list,
+            doc="Mass transfer term",
+        )
+        def eq_mass_transfer_term(self, t, p, j):
+            if j == self.boron_name_id:
+                boron_out = pyunits.convert(
+                    self.mol_Boron[t],
+                    to_units=units_meta("amount") * units_meta("length") ** -3,
+                )
+                input_rate = self.control_volume.properties_in[t].flow_mol_phase_comp[
+                    p, self.boron_name_id
+                ]
+                exit_rate = (
+                    self.control_volume.properties_out[t].flow_vol_phase[p] * boron_out
+                )
+
+                loss_rate = input_rate - exit_rate

```

Comment:
Just a question-  This constraint seems to be only for a unit conversion essentially ?

Reply:
No, this constraint is basing the loss rate off of the solution to the boron concentration inside the unit at the given pH and applies that loss rate such that we achieve that end result.

Reply:
It does include in it a unit conversion, but that is just to ensure that the units always work out regardless of how the property package was setup.